[SPARK-26811][SQL] Add capabilities to v2.Table

rdblue opened the pull request:
## What changes were proposed in this pull request?

This adds a new method, `capabilities` to `v2.Table` that returns a set of `TableCapability`. Capabilities are used to fail queries during analysis checks, `V2WriteSupportCheck`, when the table does not support operations, like truncation.

## How was this patch tested?

Existing tests for regressions, added new analysis suite, `V2WriteSupportCheckSuite`, for new capability checks.

rdblue committed:
Add capabilities to v2.Table.

SparkQA commented:
**[Test build #103156 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103156/testReport)** for PR 24012 at commit [`22f3953`](https://github.com/apache/spark/commit/22f3953d759daa6605f98c844a1c2c836c4f48d9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8610/
Test PASSed.

SparkQA commented:
**[Test build #103158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103158/testReport)** for PR 24012 at commit [`23746e7`](https://github.com/apache/spark/commit/23746e7b00da3f012ac89c831f0cebe97edf4817).

SparkQA commented:
**[Test build #103158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103158/testReport)** for PR 24012 at commit [`23746e7`](https://github.com/apache/spark/commit/23746e7b00da3f012ac89c831f0cebe97edf4817).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103158/
Test FAILed.

rdblue committed:
Add docs to TableCapability.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8613/
Test PASSed.

SparkQA commented:
**[Test build #103160 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103160/testReport)** for PR 24012 at commit [`8abfd42`](https://github.com/apache/spark/commit/8abfd42fea104b112ffe90f67cc85213bd8c79b6).

SparkQA commented:
**[Test build #103156 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103156/testReport)** for PR 24012 at commit [`22f3953`](https://github.com/apache/spark/commit/22f3953d759daa6605f98c844a1c2c836c4f48d9).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103156/
Test FAILed.

rdblue committed:
Fix Kafka source.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8614/
Test PASSed.

SparkQA commented:
**[Test build #103161 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103161/testReport)** for PR 24012 at commit [`8636867`](https://github.com/apache/spark/commit/8636867d29d4b7054a1808af613642d04b1eb21e).

SparkQA commented:
**[Test build #103160 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103160/testReport)** for PR 24012 at commit [`8abfd42`](https://github.com/apache/spark/commit/8abfd42fea104b112ffe90f67cc85213bd8c79b6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103160/
Test FAILed.

SparkQA commented:
**[Test build #103161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103161/testReport)** for PR 24012 at commit [`8636867`](https://github.com/apache/spark/commit/8636867d29d4b7054a1808af613642d04b1eb21e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103161/
Test PASSed.

rdblue commented:
@cloud-fan, can you take a look at this PR? It adds capabilities like we discussed.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/SupportsRead.java:
```diff
@@ -25,7 +25,7 @@
  * or continuous processing.
  */
-interface SupportsRead extends Table {
+public interface SupportsRead extends Table {
```
shall we remove this interface as well? We can move `newScanBuilder` to table and throw exception by default. Tables  that reports batch/stream scan capability should overwrite `newScanBuilder`

rdblue replied:
I like having this because it maintains separation between the read/write API and the catalog API. We could update the read and write API later or add a new one by adding a different read trait, without changing how catalogs and tables work. So I think it is worth keeping `SupportsRead` and `SupportsWrite`.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/Table.java:
```diff
@@ -45,4 +44,9 @@
+  /**
+   * Returns the set of capabilities for this table.
+   */
+  Set<TableCapability> capabilities();
```
I don't think we will have tons of capabilities, maybe `Array` is good enough? `Array` is also more java/scala friendly.

rdblue replied:
I don't think it's a good idea to use an array when the storage should be a set, just because it is necessary to call `asJava` when returning it.

cloud-fan commented:
LGTM except https://github.com/apache/spark/pull/24012/files#r264765864

rdblue commented:
@cloud-fan, I've rebased to pick up the changes in master introduced by the move to `CaseInsensitiveStringMap`. I think this is ready to go.

Although I see your point with returning an Array of capabilities, I think it is better to return a Set. That's how Spark uses the data and I see no reason to use the wrong kind of storage -- which we would no doubt coerce to a set -- just to avoid calling `asJava` in a few places.

SparkQA commented:
**[Test build #103455 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103455/testReport)** for PR 24012 at commit [`0d44757`](https://github.com/apache/spark/commit/0d447577a5dfbe7301b2e05b4a26a22ea43d4989).

SparkQA commented:
**[Test build #103455 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103455/testReport)** for PR 24012 at commit [`0d44757`](https://github.com/apache/spark/commit/0d447577a5dfbe7301b2e05b4a26a22ea43d4989).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103455/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8874/
Test PASSed.

cloud-fan commented:
sounds good.

rdblue committed:
Update V2WriteSupportCheckSuite for CaseInsensitiveStringMap change.

rdblue commented:
Retest this please.

SparkQA commented:
**[Test build #103510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103510/testReport)** for PR 24012 at commit [`93c77f5`](https://github.com/apache/spark/commit/93c77f5c6f2a06bc2cf8fe27fd16ff8f42a891da).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8924/
Test PASSed.

SparkQA commented:
**[Test build #103510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103510/testReport)** for PR 24012 at commit [`93c77f5`](https://github.com/apache/spark/commit/93c77f5c6f2a06bc2cf8fe27fd16ff8f42a891da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103510/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala:
```diff
@@ -58,10 +59,16 @@ abstract class FileTable(
 }
+
+object FileTable {
+  private lazy val CAPABILITIES = Set(BATCH_READ, BATCH_WRITE, TRUNCATE).asJava
```
nit: this doesn't need to be `lazy val`

rdblue replied:
I'll fix this since I need to resolve conflicts.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8957/
Test PASSed.

rdblue commented:
@cloud-fan, I've fixed the commit conflict caused by 6d22ee396959. As I noted on that commit, please do not commit non-functional changes that cause unnecessary conflicts. That problem delayed getting this work in by another day.

I've also removed `lazy` from that capabilities val.

SparkQA commented:
**[Test build #103546 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103546/testReport)** for PR 24012 at commit [`69e729e`](https://github.com/apache/spark/commit/69e729e5ec1c380e6046a0933698a08dff744395).

SparkQA commented:
**[Test build #103546 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103546/testReport)** for PR 24012 at commit [`69e729e`](https://github.com/apache/spark/commit/69e729e5ec1c380e6046a0933698a08dff744395).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103546/
Test PASSed.

rdblue commented:
@cloud-fan, tests are passing on this so it is ready for another look. Thank you!

cloud-fan commented:
thanks, merging to master!

rdblue commented:
Thank you for reviewing this, @cloud-fan!

rxin commented:
Is there plan documented on what the final API would look like? It's super confusing to have half capability via traits and half capability via enums.

cloud-fan commented:
https://github.com/apache/spark/pull/24129 is adding streaming read/write capability. Eventually we should have all the capabilities via enum.
